File: .github/scripts/coverage-badge.py
#!/usr/bin/env python3
"""Generate a shields.io endpoint badge payload from the scoverage report.

Reads the statement rate from the root scoverage aggregate report produced by
`sbt coverageAggregate` and builds a shields.io endpoint JSON payload. When
GIST_ID and GIST_TOKEN are set, the payload is pushed to the gist file named by
GIST_FILENAME; otherwise it is written locally to `coverage.json`.
"""
import glob
import json
import os
import sys
import urllib.request
import xml.etree.ElementTree as ET

OUTPUT = "coverage.json"


def find_report() -> str:
    candidates = sorted(glob.glob("target/scala-*/scoverage-report/scoverage.xml"))
    if not candidates:
        candidates = sorted(glob.glob("**/scoverage-report/scoverage.xml", recursive=True))
    if not candidates:
        print("No scoverage report found", file=sys.stderr)
        sys.exit(1)
    return candidates[0]


def statement_rate(path: str) -> float:
    root = ET.parse(path).getroot()
    return float(root.attrib["statement-rate"])


def color_for(pct: float) -> str:
    if pct >= 90:
        return "brightgreen"
    if pct >= 80:
        return "green"
    if pct >= 70:
        return "yellowgreen"
    if pct >= 60:
        return "yellow"
    if pct >= 50:
        return "orange"
    return "red"


def render(pct: float) -> str:
    payload = {
        "schemaVersion": 1,
        "label": "coverage",
        "message": f"{pct:.0f}%",
        "color": color_for(pct),
    }
    return json.dumps(payload)


def publish_to_gist(content: str) -> None:
    gist_id = os.environ["GIST_ID"]
    token = os.environ["GIST_TOKEN"]
    filename = os.environ.get("GIST_FILENAME", OUTPUT)
    body = json.dumps({"files": {filename: {"content": content}}}).encode()
    req = urllib.request.Request(
        f"https://api.github.com/gists/{gist_id}",
        data=body,
        method="PATCH",
        headers={
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "User-Agent": "coverage-badge",
        },
    )
    with urllib.request.urlopen(req) as resp:
        resp.read()
    print(f"Updated gist {gist_id} file {filename}")


def main() -> None:
    report = find_report()
    pct = statement_rate(report)
    content = render(pct)
    if os.environ.get("GIST_ID") and os.environ.get("GIST_TOKEN"):
        publish_to_gist(content)
    else:
        with open(OUTPUT, "w", encoding="utf-8") as f:
            f.write(content)
        print(f"Wrote {OUTPUT} ({pct:.0f}% from {report})")


if __name__ == "__main__":
    main()

File: .github/scripts/coverage-badge.scala
//> using scala 3.8.1
//> using dep org.scala-lang.modules::scala-xml:2.4.0
//> using dep com.lihaoyi::requests:0.9.0
//> using dep com.lihaoyi::os-lib:0.11.4

import scala.xml.XML

val Output = "coverage.json"

/** Locate the aggregated scoverage report produced by `sbt coverageAggregate`. */
def findReport(): os.Path =
  os.walk(os.pwd)
    .filter(p => p.last == "scoverage.xml" && p.segments.contains("scoverage-report"))
    .sortBy(_.toString)
    .headOption
    .getOrElse {
      System.err.println("No scoverage report found")
      sys.exit(1)
    }

def statementRate(path: os.Path): Double =
  (XML.loadFile(path.toIO) \ "@statement-rate").text.toDouble

def colorFor(pct: Double): String =
  if pct >= 90 then "brightgreen"
  else if pct >= 80 then "green"
  else if pct >= 70 then "yellowgreen"
  else if pct >= 60 then "yellow"
  else if pct >= 50 then "orange"
  else "red"

/** Render a shields.io endpoint badge payload. */
def render(pct: Double): String =
  val message = s"${pct.round}%"
  ujson.Obj(
    "schemaVersion" -> 1,
    "label"         -> "coverage",
    "message"       -> message,
    "color"         -> colorFor(pct)
  ).toString()

def publishToGist(content: String): Unit =
  val gistId   = sys.env("GIST_ID")
  val token    = sys.env("GIST_TOKEN")
  val filename = sys.env.getOrElse("GIST_FILENAME", Output)
  val body     = ujson.Obj("files" -> ujson.Obj(filename -> ujson.Obj("content" -> content)))
  val response = requests.patch(
    s"https://api.github.com/gists/$gistId",
    headers = Map(
      "Authorization" -> s"Bearer $token",
      "Accept"        -> "application/vnd.github+json",
      "User-Agent"    -> "coverage-badge"
    ),
    data = body.toString()
  )
  println(s"Updated gist $gistId file $filename (${response.statusCode})")

@main def run(): Unit =
  val report  = findReport()
  val pct     = statementRate(report)
  val content = render(pct)
  if sys.env.contains("GIST_ID") && sys.env.contains("GIST_TOKEN") then publishToGist(content)
  else
    os.write.over(os.pwd / Output, content)
    println(s"Wrote $Output (${pct.round}% from $report)")
